### PR TITLE
"Add reference here" diagnostic

### LINKED ISSUE #6814: Add "add reference" code action/diagnostics
```rust
let mut grid = 0usize;
let mut next = 92usize;
mem::swap(&mut grid, <|>next);
```

->

```rust
let mut grid = 0usize;
let mut next = 92usize;
mem::swap(&mut grid, &mut next);
```

I am not sure if we should produce a diagnostic here, but, even if we don't produce one, this should be available as a code action.

### PR DESCRIPTION
**Pull request by ivan770**
Closes #6814 

- [x] Fix incorrect suggestions in macro calls
```rust
fn main() {
      //^^^ Suggestion somehow got here
    assert!(test(true));
}

fn test(b: &bool) -> bool {
    *b
}
```

- [ ] Decide what to do with substs, if they are unknown
```rust
fn main() {
    // No suggestion here, because types are different
    test(Vec::new());
}

fn test(buf: &mut Vec<i32>) {}
```

### CODE REVIEW

File: crates/hir_ty/src/diagnostics/expr.rs [hunk 1]
```diff
@@ -155,14 +155,17 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
 
         // FIXME: Due to shortcomings in the current type system implementation, only emit this
         // diagnostic if there are no type mismatches in the containing function.
-        if self.infer.type_mismatches.iter().next().is_some() {
-            return Some(());
-        }
+        // if self.infer.type_mismatches.iter().next().is_some() {
```
Comment (ivan770) [inline crates/hir_ty/src/diagnostics/expr.rs:158]: I'm not sure if removing this check was a good decision

File: crates/hir_ty/src/diagnostics.rs [hunk 2]
```diff
@@ -318,6 +318,31 @@ impl Diagnostic for MismatchedArgCount {
     }
 }
 
+#[derive(Debug)]
+pub struct AddReferenceToArg {
+    pub file: HirFileId,
+    pub arg_expr: AstPtr<ast::Expr>,
+    pub mutability: Mutability,
+}
```
Comment (matklad) [inline crates/hir_ty/src/diagnostics.rs:326]: @jonas-schievink do we need to do anything special to make this diagnostic as experimental?

File: crates/hir_ty/src/diagnostics/expr.rs (hunk 1 above)
Comment (matklad) [reply crates/hir_ty/src/diagnostics/expr.rs:158]: Hm, I guess I am wondering a bigger thing: whey do we want to piggy-back this on the `validate_call` function? 

I think this diagnostic should be applicable anywhere where there's a type mismatch, and we should keep `validate_call` intact, to only process mismatched arg counts.

File: crates/hir_ty/src/diagnostics.rs [hunk 3]
```diff
@@ -318,6 +318,31 @@ impl Diagnostic for MismatchedArgCount {
...
+    }
+
+    fn display_source(&self) -> InFile<SyntaxNodePtr> {
+        InFile { file_id: self.file, value: self.arg_expr.clone().into() }
+    }
+
+    fn as_any(&self) -> &(dyn Any + Send + 'static) {
+        self
+    }
+}
```
Comment (jonas-schievink) [inline crates/hir_ty/src/diagnostics.rs:344]:
```suggestion

    fn is_experimental(&self) -> bool {
        true
    }
}
```

File: crates/hir_ty/src/diagnostics.rs (hunk 2 above)
Comment (jonas-schievink) [reply crates/hir_ty/src/diagnostics.rs:326]: (yeah, added the necessary code as a suggestion)

File: crates/hir_ty/src/diagnostics/expr.rs (hunk 1 above)
Comment (ivan770) [reply crates/hir_ty/src/diagnostics/expr.rs:192]: I added support for `LetStmt`, yet I still can't find any way to completely remove this diagnostic from validate_call

File: crates/hir_ty/src/diagnostics/expr.rs [hunk 4]
```diff
@@ -206,6 +244,26 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
...
 
+        let mut args_with_params = args.iter().zip(params.iter());
+
+        if is_method_call {
+            args_with_params.next();
+        }
+
+        args_with_params
+            .filter_map(|(arg, param)| {
+                let (arg, mutability) = check_missing_refs(infer, *arg, param)?;
```
Comment (matklad) [inline crates/hir_ty/src/diagnostics/expr.rs:255]: Hm, I still don't understand why we need to single out call expresisons here? It seems better to just put the logic here to `validate_body`?

This diagnostic can fire for *any* expressions, as long as `TypeMismatch` records types differing only in referenceness.

Comment (ivan770) [reply crates/hir_ty/src/diagnostics/expr.rs]: I probably got it, tests are still passing

File: crates/hir_ty/src/diagnostics/expr.rs [hunk 5]
```diff
@@ -40,7 +41,7 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
     }
 
     pub(super) fn validate_body(&mut self, db: &dyn HirDatabase) {
-        let body = db.body(self.owner.into());
+        let (body, source_map) = db.body_with_source_map(self.owner.into());
```
Comment (flodiebold) [inline crates/hir_ty/src/diagnostics/expr.rs:44]: Calling `body_with_source_map` will cause this to be recomputed whenever any change even to whitespace is done in the function, so we need to be careful to only call it when necessary (i.e. when there are actual diagnostics to emit).

Comment (ivan770) [reply crates/hir_ty/src/diagnostics/expr.rs:44]: I'll call it when we have type mismatches

**Comment by lnicola** [comment]
@ivan770 do you want to merge this as-is? It will probably need a rebase, though.

**Comment by ivan770** [comment]
> @ivan770 do you want to merge this as-is? It will probably need a rebase, though.

Hi. Yep, that would be good, but I'm not sure I'll be able to do that in July.

**Comment by ivan770** [comment]
> @ivan770 do you want to merge this as-is? It will probably need a rebase, though.

I'll close this PR in favor of #9810

**Pull request closed without merging**